src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+---
+title: "Comparing Rust Smart Contract Storage Approaches"
+pubDate: "2022-11-30"
+author: "@chadoh"
+---
+
+I  ([@chadoh](https://twitter.com/chadoh)) recently started [exploring the CosmWasm smart contract module for Cosmos](https://github.com/Web3-Builders-Alliance/ClusterOneCodeJournal.W22MTW.Chad/blob/main/standards.md) as well as [Soroban, a new smart contract platform from the Stellar team](https://github.com/stellar/sorobanathon/discussions/19).
+
+Having spent most of my time in NEAR with [near-sdk-rs](https://github.com/near/near-sdk-rs), the thing that immediately jumped out was the variety of approaches to storing on-chain data. While all these smart contract platforms use a simple key-value store, the way they wrap this key-value store in their Rust smart contracts varies wildly. This post will compare the various approaches.
+
+# Simple data
+
+As a baseline, let's see the most common approaches to storing and incrementing a number.
+
+## NEAR
+
+```rust
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::near_bindgen;
+
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct Contract {
+    counter: i8,
+}
+
+#[near_bindgen]
+impl Contract {
+    pub fn get_num(&self) -> i8 {
+        self.counter
+    }
+
+    pub fn increment(&mut self) {
+        self.counter += 1;
+    }
+}
+```
+
+If you're used to Rust, then this all looks pretty familiar. You can tell that the macros are doing all the heavy lifting.
+
+You need to add the `#[near_bindgen]` macro to a `struct`, which will make that struct the main data structure of the contract. This means all of its keys will be stored in the on-chain key-value store.
+
+NEAR also makes you import some [Borsh](https://borsh.io/) stuff to explicitly declare the encoding/decoding format of the main data structure. You can [read more about this](https://raen.dev/guide/counter/intro.html#borsh) if you're unfamiliar with the specifics of encoding and decoding. But I don't know why `near-sdk-rs` makes you be explicit in this way—every contract I've ever seen encodes and decodes with Borsh.
+
+Anyhow, the Borsh encoding means that the actual stored key will not by `counter`, as shown in the code. It will be something much smaller, like a single byte.
```

Comment:
Not really since the raw key is never used. There is a functionality to get the raw contract storage but it is limited to a contract with storage less than ~10kb which for real contracts is very easy to get to.

Reply:
Looking at the context this is also not correct in this example since the counter is inside the struct of the contract which will have a key `STATE`, which then requires being deserialized from borsh to be readable.